17.11.2025/binaryNumbers.c

```diff
@@ -0,0 +1,237 @@
+#include "binaryNumbers.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+/*
+ * Binary numbers library, implemented for 'int' type.
+ */
+
+#define BINARY_NUMBER_SIZE (sizeof(int) * 8)
+#define ARRAY_SIZE(arr) (sizeof(arr) / sizeof(arr[0]))
+
+typedef struct BinaryNumber {
+    char data[BINARY_NUMBER_SIZE];
+} BinaryNumber;
+
+/*
+ * Allocate new binary number and initialize it with zero.
+ */
+static BinaryNumber* binaryNumberAlloc()
+{
+    BinaryNumber* ptr = malloc(sizeof(*ptr));
+    if (ptr == NULL) {
+        fprintf(stderr, "Failed to allocate BinaryNumber!\n");
+        return NULL;
+    }
+    memset(ptr->data, 0, sizeof(ptr->data));
+    return ptr;
+}
+
+BinaryNumber* binaryNumberAssign(BinaryNumber* binNum, int decNum)
+{
+    if (binNum == NULL)
+        return NULL;
+
+    if (decNum == 0) {
+        memset(binNum->data, 0, sizeof(binNum->data));
+        return binNum;
+    }
+    for (int bit = BINARY_NUMBER_SIZE - 1; bit >= 0; bit--) {
+        binNum->data[bit] = ((decNum & (1 << (BINARY_NUMBER_SIZE - bit - 1))) ? 1 : 0);
```

Comment:
По-моему, если двигать наоборот, будет сильно читаемее.
```suggestion
        binNum->data[bit] = (decNum >> (BINARY_NUMBER_SIZE - bit - 1)) & 1;
```
(Проверьте, что я на правильное количество бит двигаю :))